Python/optimizer.c

```diff
@@ -591,409 +562,456 @@ translate_bytecode_to_trace(
+    _Py_CODEUNIT *target_instr = this_instr;
     uint32_t target = 0;
 
-    for (;;) {
-        target = INSTR_IP(instr, code);
-        // One for possible _DEOPT, one because _CHECK_VALIDITY itself might _DEOPT
-        max_length-=2;
-        uint32_t opcode = instr->op.code;
-        uint32_t oparg = instr->op.arg;
-
-        if (!first && instr == initial_instr) {
-            // We have looped around to the start:
-            RESERVE(1);
-            ADD_TO_TRACE(_JUMP_TO_TOP, 0, 0, 0);
-            goto done;
+    target = Py_IsNone((PyObject *)old_code)
+        ? (int)(target_instr - _Py_INTERPRETER_TRAMPOLINE_INSTRUCTIONS_PTR)
+        : INSTR_IP(target_instr, old_code);
+
+    // Rewind EXTENDED_ARG so that we see the whole thing.
+    // We must point to the first EXTENDED_ARG when deopting.
+    int oparg = tstate->interp->jit_state.prev_instr_oparg;
+    int opcode = this_instr->op.code;
+    // Failed specialization many times. Deopt!
+    if (tstate->interp->jit_state.specialize_counter >= MAX_SPECIALIZATION_TRIES) {
+        opcode = _PyOpcode_Deopt[opcode];
+    }
+    int rewind_oparg = oparg;
+    while (rewind_oparg > 255) {
+        rewind_oparg >>= 8;
+        target--;
+    }
+
+    int old_stack_level = tstate->interp->jit_state.prev_instr_stacklevel;
+
+    // Strange control-flow
+    bool has_dynamic_jump_taken = OPCODE_HAS_UNPREDICTABLE_JUMP(opcode) &&
+        (next_instr != this_instr + 1 + _PyOpcode_Caches[_PyOpcode_Deopt[opcode]]);
+    if (has_dynamic_jump_taken) {
+        DPRINTF(2, "Unsupported: dynamic jump taken\n");
+        goto unsupported;
+    }
+
+    /* Special case the first instruction,
+    * so that we can guarantee forward progress */
+    if (progress_needed && tstate->interp->jit_state.code_curr_size <= 3) {
+        if (OPCODE_HAS_EXIT(opcode) || OPCODE_HAS_DEOPT(opcode)) {
+            opcode = _PyOpcode_Deopt[opcode];
         }
+        assert(!OPCODE_HAS_EXIT(opcode));
+        assert(!OPCODE_HAS_DEOPT(opcode));
+    }
 
-        DPRINTF(2, "%d: %s(%d)\n", target, _PyOpcode_OpName[opcode], oparg);
+    bool needs_guard_ip = _PyOpcode_NeedsGuardIp[opcode];
+    DPRINTF(2, "%p %d: %s(%d) %d %d\n", old_code, target, _PyOpcode_OpName[opcode], oparg, needs_guard_ip, old_stack_level);
 
-        if (opcode == EXTENDED_ARG) {
-            instr++;
-            opcode = instr->op.code;
-            oparg = (oparg << 8) | instr->op.arg;
-            if (opcode == EXTENDED_ARG) {
-                instr--;
-                goto done;
+#ifdef Py_DEBUG
+    if (oparg > 255) {
+        assert(_Py_GetBaseCodeUnit(old_code, target).op.code == EXTENDED_ARG);
+    }
+#endif
+
+    // Skip over super instructions.
+    if (tstate->interp->jit_state.prev_instr_is_super) {
+        tstate->interp->jit_state.prev_instr_is_super = false;
+        return 1;
+    }
+
+    if (opcode == ENTER_EXECUTOR) {
+        goto full;
+    }
+
+    if (!tstate->interp->jit_state.dependencies_still_valid) {
```

Comment:
Just a thought:
If we create the executor before we start tracing, then we can check it for validity before allocating space for the uops and there will need be no need to check the jit state dependencies separately.

Reply:
I think that might make initializing tracing a lot more expensive, and will slow us down in casea cannot currently optimize.